File: Android/JavaApp/src/main/java/com/schanz/lootgen/MainApp.java
package com.schanz.lootgen;

import java.util.ArrayList;

import android.annotation.TargetApi;
import android.app.Application;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.os.Build;
import android.preference.PreferenceManager;

import com.schanz.lootgen.Constants.CreationType;
import com.schanz.lootgen.data.Creation;

public class MainApp extends Application {

    // Static methods and variables.
    private static MainApp sSingleton;

    private static final String STORED_PREFERENCE_USERNAME = "storedUsername";
    private static final String STORED_PREFERENCE_PASSWORD = "storedPassword";

    public static final int ACHIEVEMENT_HIT_RANDOM = 10;

    public static int sHitsRandom = 0;

    public static MainApp getInstance() {
        return sSingleton;
    }

    @TargetApi(Build.VERSION_CODES.GINGERBREAD)
    public static void storeUsernamePreference(String lastUsername) {
        SharedPreferences preferences =
                PreferenceManager.getDefaultSharedPreferences(sSingleton);
        Editor preferencesEditor = preferences.edit();
        preferencesEditor.putString(STORED_PREFERENCE_USERNAME, lastUsername);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.GINGERBREAD) {
            preferencesEditor.apply();
        } else {
            preferencesEditor.commit();
        }
    }

    @TargetApi(Build.VERSION_CODES.GINGERBREAD)
    public static void storePasswordPreference(String lastPassword) {
        SharedPreferences preferences =
                PreferenceManager.getDefaultSharedPreferences(sSingleton);
        Editor preferencesEditor = preferences.edit();
        preferencesEditor.putString(STORED_PREFERENCE_PASSWORD, lastPassword);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.GINGERBREAD) {
            preferencesEditor.apply();
        } else {
            preferencesEditor.commit();
        }
    }

    @TargetApi(Build.VERSION_CODES.GINGERBREAD)
    public static void removeStoredPreference(String key) {
        SharedPreferences preferences =
                PreferenceManager.getDefaultSharedPreferences(sSingleton);
        Editor preferencesEditor = preferences.edit();
        preferencesEditor.remove(key);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.GINGERBREAD) {
            preferencesEditor.apply();
        } else {
            preferencesEditor.commit();
        }
    }

    public static String getStoredUsername() {
        SharedPreferences preferences =
                PreferenceManager.getDefaultSharedPreferences(sSingleton);
        return preferences.getString(STORED_PREFERENCE_USERNAME, null);
    }

    public static String getStoredPassword() {
        SharedPreferences preferences =
                PreferenceManager.getDefaultSharedPreferences(sSingleton);
        return preferences.getString(STORED_PREFERENCE_PASSWORD, null);
    }

    // Non-static methods and variables.
    @Override
    public void onCreate() {
        super.onCreate();
        sSingleton = this;
    }

    private ArrayList<Creation> mCreations = new ArrayList<Creation>();

    private Creation mCurrentCreation = new Creation();

    public Creation getCurrentCreation() {
        return mCurrentCreation;
    }

    public void setCurrentCreation(Creation currentCreation) {
        mCurrentCreation = currentCreation;
    }

    /**
     * Sets the {@code CreationType} of the current global {@code Creation} object.
     *
     * @return The current object's {@code CreationType} or null..
     */
    public void setCurrentCreationType(CreationType creationType) {
        mCurrentCreation.setCreationType(creationType);
    }

    /**
     * Gets the {@code CreationType} of the current global {@code Creation} object.
     *
     * @return The current object's {@code CreationType} or null.
     */
    public CreationType getCurrentCreationType() {
        return mCurrentCreation.getCreationType();
    }

    /**
     * Checks if the given <b>drawable</b> resource name exists within this application.
     *
     * @param resourceName The name of the image resource.
     * @return {@code true} if the resource exists, {@code false} if it does not.
     */
    public boolean doesResourceExist(String resourceName) {
        int exists = getResources().getIdentifier(resourceName, "drawable", getPackageName());
        return exists != 0 ? true : false;
    }

    public ArrayList<Creation> getCreations() {
        return mCreations;
    }

    public void setCreations(ArrayList<Creation> creations) {
        mCreations = creations;
    }
}

File: Android/KtApp/src/main/java/com/schanz/ktapp/MainApp.kt
package com.schanz.ktapp

import android.annotation.TargetApi
import android.app.Application
import android.os.Build
import androidx.preference.PreferenceManager
import com.schanz.ktapp.data.Creation
import java.util.*

class MainApp : Application() {

    // Declare "static" methods and variables in the companion object.
    companion object {
        // Special prefixes are not idiomatic in Kotlin. Refer to the official style guide:
        // https://developer.android.com/kotlin/style-guide#naming_2
        lateinit var sSingleton: MainApp

        private const val STORED_PREFERENCE_USERNAME = "storedUsername"
        private const val STORED_PREFERENCE_PASSWORD = "storedPassword"

        const val ACHIEVEMENT_HIT_RANDOM = 10

        @JvmField
        var sHitsRandom = 0

        @JvmStatic // For interoperability with existing Java code.
        fun getInstance(): MainApp {
            return sSingleton
        }

        @TargetApi(Build.VERSION_CODES.GINGERBREAD)
        fun storeUsernamePreference(lastUsername: String?) {
            val preferences = PreferenceManager.getDefaultSharedPreferences(sSingleton)
            val preferencesEditor = preferences.edit()
            preferencesEditor.putString(STORED_PREFERENCE_USERNAME, lastUsername)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.GINGERBREAD) {
                preferencesEditor.apply()
            } else {
                preferencesEditor.commit()
            }
        }

        @TargetApi(Build.VERSION_CODES.GINGERBREAD)
        fun storePasswordPreference(lastPassword: String?) {
            val preferences = PreferenceManager.getDefaultSharedPreferences(sSingleton)
            val preferencesEditor = preferences.edit()
            preferencesEditor.putString(STORED_PREFERENCE_PASSWORD, lastPassword)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.GINGERBREAD) {
                preferencesEditor.apply()
            } else {
                preferencesEditor.commit()
            }
        }

        @TargetApi(Build.VERSION_CODES.GINGERBREAD)
        fun removeStoredPreference(key: String?) {
            val preferences = PreferenceManager.getDefaultSharedPreferences(sSingleton)
            val preferencesEditor = preferences.edit()
            preferencesEditor.remove(key)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.GINGERBREAD) {
                preferencesEditor.apply()
            } else {
                preferencesEditor.commit()
            }
        }

        val storedUsername: String?
            get() {
                val preferences = PreferenceManager.getDefaultSharedPreferences(sSingleton)
                return preferences.getString(STORED_PREFERENCE_USERNAME, null)
            }

        val storedPassword: String?
            get() {
                val preferences = PreferenceManager.getDefaultSharedPreferences(sSingleton)
                return preferences.getString(STORED_PREFERENCE_PASSWORD, null)
            }
    }

    // Non-static methods and variables.
    override fun onCreate() {
        super.onCreate()
        sSingleton = this
    }

    var creations: List<Creation> = ArrayList<Creation>()
    var currentCreation: Creation = Creation()

    /**
     * Getter and setter for the `CreationType` of the current global `Creation` object.
     *
     * @return The current object's `CreationType`, or null.
     */
    var currentCreationType: Constants.CreationType?
        get() = currentCreation.creationType
        set(creationType) {
            currentCreation.creationType = creationType
        }

    /**
     * Checks if the given **drawable** resource name exists within this application.
     *
     * @param resourceName The name of the image resource.
     * @return `true` if the resource exists, `false` if it does not.
     */
    fun doesResourceExist(resourceName: String?): Boolean {
        val exists = resources.getIdentifier(resourceName, "drawable", packageName)
        return exists != 0
    }
}
